src/codegen/constant-propagation.lisp

```diff
@@ -0,0 +1,129 @@
+                     (make-node-let
+                      :type (node-type node)
+                      :bindings nonconstant-bindings
+                      :subexpr (funcall *traverse* (node-let-subexpr node) inner-constant-bindings))
+                     (funcall *traverse* (node-let-subexpr node) inner-constant-bindings)))))
+
+           (propagate-constants-node-lisp (node constant-bindings)
+             (declare (type node-lisp node))
+             (let* ((new-lisp-vars nil)
+                    (let-bindings
+                      (loop :for (lisp-var . coalton-var) :in (node-lisp-vars node)
+                            :for constant-value := (constant-var-value coalton-var constant-bindings :no-error t)
+                            :if constant-value
+                              :collect (let ((new-coalton-var (gentemp (symbol-name coalton-var))))
+                                         (push (cons lisp-var new-coalton-var) new-lisp-vars)
+                                         (cons new-coalton-var constant-value))
+                            :else
+                              :do (push (cons lisp-var coalton-var) new-lisp-vars))))
+               (if let-bindings
+                   (make-node-let
+                    :type (node-type node)
+                    :bindings let-bindings
+                    :subexpr (make-node-lisp
+                              :type (node-type node)
+                              :vars new-lisp-vars
+                              :form (node-lisp-form node)))
+                   (make-node-lisp
+                    :type (node-type node)
+                    :vars new-lisp-vars
+                    :form (node-lisp-form node)))))
+
+           (direct-application-better-infer-types (node constant-bindings)
+             ;; FIXME: Can we do something similar for non-direct APPLICATION?
+             (let ((constant-propagated-rands
+                     (mapcar (lambda (rand)
+                               (funcall *traverse* rand constant-bindings))
+                             (node-direct-application-rands node))))
+               (make-node-direct-application
+                :type (node-type node)
+                :rator-type (tc:make-function-type*
+                             (mapcar #'node-type constant-propagated-rands)
+                             (tc:function-return-type (node-type node)))
+                :rator (node-direct-application-rator node)
+                :rands constant-propagated-rands))))
+
+    (let ((*constant-bindings* nil))
+      (traverse
+       node
+       (list
+        (action (:after node-variable) #'propagate-constants-node-variable)
+        (action (:traverse node-let) #'propagate-constants-node-let)
+        (action (:after node-lisp) #'propagate-constants-node-lisp)
+        (action (:after node-direct-application) #'direct-application-better-infer-types)
+        (make-traverse-let-action-skipping-cons-bindings))
```

Comment:
```suggestion
        (action (:after node-direct-application) #'direct-application-better-infer-types))
```
I can't think of a reason why the constant propagation should need to skip `let`-bindings to `Cons` applications (this special action was a very particular fix to make sure the inliner wouldn't break recursive `Cons` bindings)